### DISCUSSION TITLE
Third Party Compression compatibility

### DISCUSSION THREAD

**Discussion by vishnucheerakoda** [klauspost/compress#800] [General]
Hi all,
 I am working on ZSTD Compression/Decompression algorithms. I have used Facebook Decompressor as well as their Educational decoder to validate RTL compression algorithm. Found few issues ,those are stated below. 

 #1 My RTL compressed data is not getting decompressed with latest Facebook ZSTD Decompressor, where as their 
    Educational decoder is able to do it.
    Note: Production is able to decode it but the data is wrong.(something is going wrong when it is a 4 stream encoded, 
    sequences are found okay)

#2 Educational decoder fails for few blocks of webster and samba due to a corner case:
  The case is all the 256 literals are having same weight(let's say 1)
  So the Huffman codes would be like 0x1000
  It is not able to find the end since there is no overflow on the stream.

Files : 
[1st_case.txt.zip](https://github.com/klauspost/compress/files/11145810/1st_case.txt.zip)

**Comment by klauspost** [comment]
The output produces [`corruption detected: stream overrun 4`](https://github.com/klauspost/compress/blob/5d754d7eb416623d7f93345a31a0589872a31f03/huff0/decompress_generic.go#L170).

This indicates that the huff0 stream produces more output than it should.

```
2023/04/04 09:56:44 Frame: Dict: 0 FrameContentSize: 32768 singleseg: true window: 32768 crc: false
2023/04/04 09:56:44 decoding new block
2023/04/04 09:56:44 Data size on stream: 16294
2023/04/04 09:56:44 History trimmed: 0 decoded already: 0
2023/04/04 09:56:44 literals type: literalsBlockCompressed litRegenSize: 6979 litCompSize: 4473 sizeFormat: 2 4X: true
2023/04/04 09:56:44 huff table input: 4473 CRC: 666845455
2023/04/04 09:56:44 decoding compressed literals: corruption detected: stream overrun 4
```

Huff0 Stream 0 decodes to at least 1746 bytes.

**Reply by vishnucheerakoda** [reply]
Hi,
  Thank you for your valuable response. i would like to get more clarity on the 1st question
this is exactly what i am getting when i use the educational decoder/our own decoder.
 
Stream 0 size is 1110(in decimal) and it generates 1752 literals
Stream 1 size is 1112(in decimal) and it generates 1752 literals
Stream 2 size is 1105(in decimal) and it generates 1744 literals
Stream 3 size is 1093(in decimal) and it generates 1731 literals
So the sum of these are 6979 which is exactly matching.
I am unable to figure out where it is going wrong !

Could you please try to get the Huffman codes generated for the literals?
I have attached the text file which has Huffman codes for the literals.


[lit_len_code.txt](https://github.com/klauspost/compress/files/11147392/lit_len_code.txt)

Thanks

**Reply by klauspost** [reply]
This is the decoded table as I see it here:
```
09	bits: 6 code:008
0a	bits: 5 code:009
20	bits: 4 code:009
21	bits: 8 code:006
26	bits: b code:000
27	bits: 7 code:009
2c	bits: 6 code:009
2d	bits: 8 code:007
2e	bits: 6 code:00a
3a	bits: 7 code:00a
3b	bits: 8 code:008
3f	bits: 7 code:00b
41	bits: 7 code:00c
42	bits: 9 code:004
43	bits: 8 code:009
44	bits: 9 code:005
45	bits: 8 code:00a
46	bits: 9 code:006
47	bits: 9 code:007
48	bits: 9 code:008
49	bits: 6 code:00b
4a	bits: 9 code:009
4b	bits: b code:001
4c	bits: 8 code:00b
4d	bits: 8 code:00c
4e	bits: 8 code:00d
4f	bits: 7 code:00d
50	bits: a code:003
51	bits: b code:002
52	bits: 8 code:00e
53	bits: 8 code:00f
54	bits: 8 code:010
55	bits: 9 code:00a
56	bits: b code:003
57	bits: 8 code:011
59	bits: a code:004
5b	bits: b code:004
5d	bits: b code:005
61	bits: 4 code:00a
62	bits: 6 code:00c
63	bits: 6 code:00d
64	bits: 5 code:00a
65	bits: 4 code:00b
66	bits: 6 code:00e
67	bits: 6 code:00f
68	bits: 5 code:00b
69	bits: 4 code:00c
6a	bits: a code:005
6b	bits: 7 code:00e
6c	bits: 5 code:00c
6d	bits: 5 code:00d
6e	bits: 5 code:00e
6f	bits: 4 code:00d
70	bits: 6 code:010
71	bits: 9 code:00b
72	bits: 5 code:00f
73	bits: 4 code:00e
74	bits: 4 code:00f
75	bits: 5 code:010
76	bits: 7 code:00f
77	bits: 6 code:011
78	bits: a code:006
79	bits: 5 code:011
7a	bits: a code:007
```

It is the same as you posted.

**Reply by klauspost** [reply]
Stream 0 decoding:

[decoded_stream0.txt](https://github.com/klauspost/compress/files/11147802/decoded_stream0.txt)

**Reply by vishnucheerakoda** [reply]
Yes. this is matching. are you seeing the corruption on 1st stream itself ? , could you please give some more hints ?
could you please provide me the encoded stream dump(for all the 4 streams) using above Huffman codes if possible? that would be really helpful to compare.
**Please consider below literal counts to split the attached literal files into 4 streams.**
stream 0 : no of literals to be encoded is: 1752 
stream 1 : no of literals to be encoded is: 1752 
stream 2 : no of literals to be encoded is: 1744 
stream 3 : no of literals to be encoded is: 1731
[raw_literals.txt](https://github.com/klauspost/compress/files/11148030/raw_literals.txt)

**Reply by vishnucheerakoda** [reply]
> Stream 0 decoding:
> 
> [decoded_stream0.txt](https://github.com/klauspost/compress/files/11147802/decoded_stream0.txt)

I could see 1745 literals decoded in the text file. but actually 7 more literals are to be decoded. is the stream ended here ?

**Reply by klauspost** [reply]
The output size is 6979, so stream 0 should be `(6979 + 3)/4` = 1745 bytes.

**Reply by klauspost** [reply]
You cannot just split the streams as you like.

**Reply by klauspost** [reply]
~(though I cannot find that in the spec)~

**Reply by klauspost** [reply]
Here it is: https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md#jump-table

> The decompressed size of each stream is equal to (Regenerated_Size+3)/4, except for the last stream which may be up to 3 bytes smaller, to reach a total decompressed size as specified in `Regenerated_Size`.

**Reply by vishnucheerakoda** [reply]
Thanks for the valuable information and quick response. I have modified and it started working now.

I have one more question regarding the probability generation part.

There are 2 methods to calculate the probabilities of symbols, it switches to 2nd method if the1st one fails. have you ever noticed 1st method failing frequently ? they have stated that **it only needs for corner cases** . I am following 1st method . what would be the impact on CR ?
https://github.com/facebook/zstd/blob/dev/lib/compress/fse_compress.c
line 503

**Reply by klauspost** [reply]
I have the [same fallback](https://github.com/klauspost/compress/blob/master/zstd/fse_encoder.go#L259), and I haven't reviewed it.

**Reply by klauspost** [reply]
Running fuzz tests shows it is getting hit:

![image](https://user-images.githubusercontent.com/5663952/229871144-c20df8ff-179c-413e-83a5-a994acd662f3.png)

**Reply by vishnucheerakoda** [reply]
Thanks for the information. how much improvement on CR we can expect for **ZSTD 32KB block size** based processing with respect to GZIP/ZLIB on benchmark suits (silesia,calgary,cantrbury,large_cantrbury) ? I am getting almost same.

**Reply by klauspost** [reply]
The base format isn't that different from deflate, but it is much faster to decompress. The biggest change is from the larger window size.

Repeat matches help a little on certain content, mostly machine generated data.

Comparing the fastest deflate in Go for CSV input it goes from 846,471,964@221Mb/s -> 641,319,332@382MB/s for CSV content - source size 3,325,605,752.

You can see my benchmarks here: https://docs.google.com/spreadsheets/d/1nuNE2nPfuINCZJRMt6wFWhKpToF95I47XjSsc-1rbPQ/edit?usp=sharing

You can download my test sets from here: https://files.klauspost.com/compress/

I don't care too much about silesia, calgary, cantrbury, large_cantrbury, since they don't really represent common data. I mostly judge speed against CSV, JSON, Log data and various machine generated data.

**Reply by klauspost** [reply]
Other than that it mostly comes down to a good search algorithm. For the fastest I use a 6 byte hash. Similar to zstd I use two hash tables in default mode, one for long (8 bytes) and one for short matches (4 bytes) with preference to the long. No chaining, but when a short match is found it will try to find a better a byte further forward.

"Better" has bigger hash tables and keeps two last matches per value and does more thorough check of candidates.

"Best" does a pretty exhaustive search with chain depth 1 and up to two bytes ahead, as well as some end-of-match trickery. I tried deeper chaining, but it didn't give much benefit ([my latest attempt](https://github.com/klauspost/compress/commit/e5f7a4d971837d91c01c8fa7d8602b961f87f02b))

I never consider matches of 3 bytes.

**Reply by vishnucheerakoda** [reply]
Okay. ZSTD gets better when the block size is more and the history is more. for hardware we can't afford to have more than 128KB history due to area limitation. how much history does it supports ?(not as per the spec which is really huge). how do we restrict the history window size when we compress a file with the application ? is there any parameter which can be passed ? have you tried systolic array based search algorithm ?

**Reply by klauspost** [reply]
> The minimum Window_Size is 1 KB. The maximum Window_Size is (1<<41) + 7*(1<<38) bytes, which is 3.75 TB.

Above 4MB the gains aren't that big, unless you have very special data. My implementation uses 4, 8, 16, 32MB windows for the different levels. The zstd spec recommonds:

> For improved interoperability, it's recommended for decoders to support Window_Size of up to 8 MB, and it's recommended for encoders to not generate frame requiring Window_Size larger than 8 MB. It's merely a recommendation though, decoders are free to support larger or lower limits, depending on local limitations.

128K will be a limitation. Will it be 128K independent blocks or 128K Window size? Independent blocks will suffer much more than a constant 128K Window, but where backreferences aren't reset.

I tested a hardware compression implementation with 128K independent blocks, and at its highest compression level it couldn't even match zstd at level 1 in terms of compression. It was ~equivalent to my deflate at level 3 on "nyc-taxi".

**Reply by vishnucheerakoda** [reply]
It won't be independent 128KB . it can slide between blocks till 128KB. hardware compression has lot of limitations to take decisions based on the data , eg : repeat FSE table, use predefined table etc.

**Reply by klauspost** [reply]
Predefined is actually a bigger deal than I expected. Here is a comparison, using defaults, using 128K window and only predefined tables and using 128K but with dynamic tables on `nyc-taxi-data-10M.csv`:

Level | Defaults | 128K + Predef | 128K
-- | -- | -- | --
1 | 641,319,332 | 826,919,668 | 709,479,324
2 | 588,976,126 | 818,387,598 | 690,662,612
3 | 517,666,647 | 753,742,717 | 650,892,393
4 | 459,777,201 | 712,467,733 | 620,587,433

Gzip reference: 1: 846,471,964 - 5: 767,340,514 - 9: 709,819,738
Zstandard reference: 1: 687,399,637 - 3: 598,514,411 - 9: 517,554,797

**Reply by klauspost** [reply]
For reference, the hw implementation I tested output 791,xxx,xxx bytes. But it did all entropy coding (huff0+fse) in software.

**Reply by vishnucheerakoda** [reply]
Hi, I would like to know how did you get the levels of ZSTD compression ? did you use lz bench to get that ? if so, the block size which we mention in the command (lzbench -b128 -v -ezstd file) will restrict the distance <=128KB ?

**Reply by klauspost** [reply]
@vishnucheerakoda I used my own library for these numbers (it has a switch to force predefined tables). My library only has 4 levels. 

Block size was 128K for all. *Window* size was 128K.

The "Zstandard reference" was using the commandline `zstd -1`, etc.

**Comment by klauspost** [comment]
Do you have a sample of the huff0 case? I don't think it is possible.

My encoder will never generate that case since it will reject this as incompressible after doing the histogram.

In fact I cannot encode this, since the table is incompressible as FSE and cannot be stored as raw values.

The first check is [`maxCount < (len(in)>>7)`](https://github.com/klauspost/compress/blob/34dac29a9a3fa3bc6d1bd7735695c4d0e9486754/huff0/compress.go#L77)

This means I will reject trying to compress if no histogram entry is above `len(input)/128`. 

The second check is if the FSE is incompressible and [`maxSymbolValue > (256 - 128)`](https://github.com/klauspost/compress/blob/910cf1695b98e4523464bb46457c975f3850d2cd/huff0/huff0.go#L234).

Finally, I reject if actual output + table [doesn't give a reasonable improvement](https://github.com/klauspost/compress/blob/34dac29a9a3fa3bc6d1bd7735695c4d0e9486754/huff0/compress.go#L155) after compression. This is tweakable, but for zstd it is fixed at `wantSize -= wantSize >> 4` for now.

This is to avoid the decompression penalty when improvement is marginal.

**Reply by klauspost** [reply]
In short, I cannot see how you can generate a valid table for the case you describe since the FSE input is a single value repeated, so if you could provide a sample input I can check that.

**Reply by vishnucheerakoda** [reply]
Lets say all the literals are having same frequency. so the Huffman code length would be 8 for all the literals(as you say uncompressible, RTL doesn't have that intelligence). so the probability values for a TableLog of 6 would be {0,64}. which indicates symbol 1 is occupying all the states in the fse table. as per spec this is valid. so the huffman code encoded stream would be just 0x1000 . **0001|SOS|000000|even state lit0|000000|odd state lit1|** , rest all would be zero. so total stream size would be only 2 bytes. as per educational decoder it will be waiting for overflow , but here it wouldn't happen even if you consume 256 literals.

**Reply by klauspost** [reply]
> which indicates symbol 1 is occupying all the states in the fse table. as per spec this is valid

No. The FSE code explicitly disallow for single value:

> Note that there must be two or more symbols with nonzero probability. [source](https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md#fse-table-description)

Since all weights have the same value it cannot be represented as FSE.

**Reply by vishnucheerakoda** [reply]
This makes sense!!! Thank you!!!
